### PR TITLE
Bluetooth: host: Fix array overrun of client features bitset

### LINKED ISSUE #32406: [Coverity CID :218737] Out-of-bounds write in subsys/bluetooth/host/gatt.c
Static code scan issues found in file:

https://github.com/zephyrproject-rtos/zephyr/tree/fe7c2efca800a0cf1bccf23aefe08b3c4beb88bf/subsys/bluetooth/host/gatt.c#L504

Category: Memory - corruptions
Function: `cf_set_value`
Component: Bluetooth
CID: [218737](https://scan9.coverity.com/reports.htm#v29726/p12996/mergedDefectId=218737)

Details:

```
493         uint16_t i;
494         uint8_t last_byte = CF_BYTE_LAST;
495         uint8_t last_bit = CF_BIT_LAST;
496    
497         /* Validate the bits */
498         for (i = 0U; i < len && i <= last_byte; i++) {
>>>     CID 218737:    (OVERRUN)
>>>     Overrunning array "cfg->data" of 1 bytes at byte offset 2 using index "i" (which evaluates to 2).
499             uint8_t chg_bits = value[i] ^ cfg->data[i];
500             uint8_t bit;
501    
502             for (bit = 0U; bit <= last_bit; bit++) {
503                 /* A client shall never clear a bit it has set */
504                 if ((BIT(bit) & chg_bits) &&
507                 }
508             }
509         }
510    
511         /* Set the bits for each octect */
512         for (i = 0U; i < len && i < last_byte; i++) {
>>>     CID 218737:    (OVERRUN)
>>>     Overrunning array "cfg->data" of 1 bytes at byte offset 1 using index "i" (which evaluates to 1).
513             cfg->data[i] |= value[i] & (BIT(last_bit + 1) - 1);
514             BT_DBG("byte %u: data 0x%02x value 0x%02x", i, cfg->data[i],
515                    value[i]);
516         }
517    
518         return true;
498         for (i = 0U; i < len && i <= last_byte; i++) {
499             uint8_t chg_bits = value[i] ^ cfg->data[i];
500             uint8_t bit;
501    
502             for (bit = 0U; bit <= last_bit; bit++) {
503                 /* A client shall never clear a bit it has set */
>>>     CID 218737:    (OVERRUN)
>>>     Overrunning array "cfg->data" of 1 bytes at byte offset 2 using index "i" (which evaluates to 2).
504                 if ((BIT(bit) & chg_bits) &&
505                     (BIT(bit) & cfg->data[i])) {
506                     return false;
507                 }
508             }
509         }
```

Please fix or provide comments in coverity using the link:

https://scan9.coverity.com/reports.htm#v32951/p12996.

Note: This issue was created automatically. Priority was set based on classification
of the file affected and the impact field in coverity. Assignees were set using the CODEOWNERS file.

### PR DESCRIPTION
**Pull request by joerchan**
Fix array overrun of client features bitset, where CF_BYTE_LAST was
incorrectly defined as 2 using module instead of divide.
Also fix additional bit-logic when number of bits would exceed a byte:
 - Inner for loop iterates from 0 to number of bits on a single byte.
 - Same bitmask used for all bytes.

Fixes: #32406

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -425,7 +425,8 @@ enum {
 #define CF_BIT_NOTIFY_MULTI	2
 #define CF_BIT_LAST		CF_BIT_NOTIFY_MULTI
 
-#define CF_BYTE_LAST		(CF_BIT_LAST % 8)
+#define CF_NUM_BITS             (CF_BIT_LAST + 1)
```
Comment (jhedberg) [inline subsys/bluetooth/host/gatt.c:428]: I don't see this used anywhere. Should it be?

Comment (joerchan) [reply subsys/bluetooth/host/gatt.c:428]: I saw that too, updated with where I wanted to use it :)

File: subsys/bluetooth/host/gatt.c [hunk 2]
```diff
@@ -491,26 +492,24 @@ static ssize_t cf_read(struct bt_conn *conn, const struct bt_gatt_attr *attr,
...
 
 	/* Validate the bits */
-	for (i = 0U; i < len && i <= last_byte; i++) {
-		uint8_t chg_bits = value[i] ^ cfg->data[i];
-		uint8_t bit;
-
-		for (bit = 0U; bit <= last_bit; bit++) {
+	for (i = 0U; i <= CF_BIT_LAST && (i / 8) < len; i++) {
+		if ((cfg->data[i / 8] & (i % 8)) &&
+		    !(value[i / 8] & (i % 8))) {
```
Comment (trond-snekvik) [inline subsys/bluetooth/host/gatt.c:499]: `value[i / 8] & BIT(i % 8)`?

Comment (joerchan) [reply subsys/bluetooth/host/gatt.c]: Yes, thank you.

**Review by jhedberg — APPROVED** [review]

**Comment by joerchan** [comment]
Tested the function in isolation with a couple of different inputs and a longer byte array i.e more bits.

@Vudentz Do you know if this feature is covered by EDTT?

**Review by trond-snekvik — APPROVED** [review]

**Pull request merged**